// @glif/filecoin-number/lib/Converter.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import bent from 'bent';
import { BigNumber } from 'bignumber.js';
import { FilecoinNumber } from './FilecoinNumber';
export var Converter = /*#__PURE__*/function () {
  function Converter(currency, options) {
    _classCallCheck(this, Converter);

    _defineProperty(this, "currency", void 0);

    _defineProperty(this, "apiKey", void 0);

    _defineProperty(this, "apiURL", void 0);

    _defineProperty(this, "rate", void 0);

    if (!currency) throw new Error('No currency passed.');
    this.currency = currency;
    this.apiKey = (options === null || options === void 0 ? void 0 : options.apiKey) || '';
    this.apiURL = (options === null || options === void 0 ? void 0 : options.apiURL) || 'https://pro-api.coinmarketcap.com/';
    this.rate = null;
  }

  _createClass(Converter, [{
    key: "cacheConversionRate",
    value: function () {
      var _cacheConversionRate = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.t0 = BigNumber;
                _context.next = 3;
                return this.convert(1, 'FIL', this.currency);

              case 3:
                _context.t1 = _context.sent;
                this.rate = new _context.t0(_context.t1);

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function cacheConversionRate() {
        return _cacheConversionRate.apply(this, arguments);
      }

      return cacheConversionRate;
    }()
  }, {
    key: "toFIL",
    value: function toFIL(amount) {
      if (!amount) return this.toFIL('0');

      if (!this.rate) {
        throw new Error('Call cacheConversionRate() to get the conversion rate before calling .toFIL.');
      }

      if (typeof amount === 'string' || typeof amount === 'number' || amount instanceof BigNumber) {
        var filAmount = new BigNumber(amount).dividedBy(this.rate);
        return new FilecoinNumber(filAmount, 'fil');
      }

      throw new Error('Amount passed must be a Number, String, or an instanceof BigNumber');
    }
  }, {
    key: "fromFIL",
    value: function fromFIL(amount) {
      if (!amount) return this.fromFIL('0');
      if (!this.rate) throw new Error('Call cacheConversionRate() to get the conversion rate before calling .fromFIL.');

      if (typeof amount === 'string' || typeof amount === 'number' || amount instanceof BigNumber) {
        return new BigNumber(amount).multipliedBy(this.rate);
      }

      throw new Error('Amount passed must be a Number, String, or an instanceof BigNumber.');
    }
  }, {
    key: "convert",
    value: function () {
      var _convert = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(amount, from, to) {
        var get, res;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                get = bent('GET', 'json', {
                  'X-CMC_PRO_API_KEY': this.apiKey
                });
                _context2.next = 3;
                return get("".concat(this.apiURL, "/v1/tools/price-conversion?symbol=").concat(from, "&amount=").concat(amount, "&convert=").concat(to));

              case 3:
                res = _context2.sent;

                if (!(!res.data || !res.data.quote || !res.data.quote[to])) {
                  _context2.next = 6;
                  break;
                }

                throw new Error('No conversion price found.');

              case 6:
                return _context2.abrupt("return", res.data.quote[to].price);

              case 7:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function convert(_x, _x2, _x3) {
        return _convert.apply(this, arguments);
      }

      return convert;
    }()
  }, {
    key: "getCachedConversionRate",
    value: function getCachedConversionRate() {
      return this.rate;
    }
  }]);

  return Converter;
}();
