// @glif/filecoin-wallet-provider/lib/providers/local-node-provider.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import LotusRpcEngine from '@glif/filecoin-rpc-client';
export var LocalNodeProvider = /*#__PURE__*/function () {
  function LocalNodeProvider(options) {
    _classCallCheck(this, LocalNodeProvider);

    _defineProperty(this, "apiAddress", void 0);

    _defineProperty(this, "token", void 0);

    _defineProperty(this, "jsonRpcEngine", void 0);

    this.apiAddress = options.apiAddress;
    this.token = options.token;
    this.jsonRpcEngine = new LotusRpcEngine({
      apiAddress: options.apiAddress,
      token: options.token
    });
  }

  _createClass(LocalNodeProvider, [{
    key: "newAccount",
    value: function newAccount() {
      return this.jsonRpcEngine.request('WalletNew', 'secp256k1');
    }
  }, {
    key: "getAccounts",
    value: function getAccounts() {
      return this.jsonRpcEngine.request('WalletList');
    }
  }, {
    key: "sign",
    value: function sign(from, message) {
      return this.jsonRpcEngine.request('WalletSignMessage', from, message);
    }
  }]);

  return LocalNodeProvider;
}();
