// send-crypto/build/main/lib/retry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.retryNTimes = exports.fallback = exports.onlyTestnet = exports.onlyMainnet = exports.extractError = exports.sleep = void 0;
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
exports.sleep = sleep;
const extractError = (error) => {
    if (error && typeof error === "object") {
        if (error.response) {
            return (0, exports.extractError)(error.response);
        }
        if (error.error) {
            return (0, exports.extractError)(error.error);
        }
        if (error.message) {
            return (0, exports.extractError)(error.message);
        }
        if (error.data) {
            return (0, exports.extractError)(error.data);
        }
        if (error.context) {
            return (0, exports.extractError)(error.context);
        }
        if (error.statusText) {
            return (0, exports.extractError)(error.statusText);
        }
        try {
            return JSON.stringify(error);
        }
        catch (error) {
            // Ignore JSON error
        }
    }
    // Remove `Error: ` prefix.
    try {
        if (typeof error === "string") {
            if (error.slice(0, 7).toLowerCase() === "Error: ") {
                // tslint:disable-next-line: no-parameter-reassignment
                error = error.slice(7);
            }
            return error;
        }
        return JSON.stringify(error);
    }
    catch (error) {
        // Ignore JSON error
    }
    return String(error);
};
exports.extractError = extractError;
const onlyMainnet = (x, testnet) => testnet ? undefined : x;
exports.onlyMainnet = onlyMainnet;
const onlyTestnet = (x, testnet) => testnet ? x : undefined;
exports.onlyTestnet = onlyTestnet;
const fallback = async (fallbacks) => {
    let firstError;
    for (const fn of fallbacks) {
        if (!fn) {
            continue;
        }
        try {
            return await fn();
        }
        catch (error) {
            firstError = firstError || error;
        }
    }
    throw firstError || new Error("No result returned");
};
exports.fallback = fallback;
const retryNTimes = async (fnCall, retries) => {
    let returnError;
    for (let i = 0; i < retries; i++) {
        try {
            return await fnCall();
        }
        catch (error) {
            if (String(error).match(/timeout of .* exceeded/)) {
                returnError = error;
            }
            else {
                const errorMessage = (0, exports.extractError)(error);
                if (errorMessage) {
                    // tslint:disable-next-line: no-object-mutation
                    error.message += ` (${errorMessage})`;
                }
                throw error;
            }
        }
        await (0, exports.sleep)(500);
    }
    throw returnError;
};
exports.retryNTimes = retryNTimes;
