// send-crypto/build/module/common/apis/bitcoinDotCom.js
import axios from "axios";
import { fixUTXO, fixUTXOs, fixValue, sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
const endpoint = (testnet) => testnet ? "https://trest.bitcoin.com/v2/" : "https://rest.bitcoin.com/v2/";
const endpointV2 = (testnet) => testnet
    ? "https://explorer-tbch.api.bitcoin.com/tbch/v1"
    : "https://explorer.api.bitcoin.com/bch/v1";
const fetchUTXO = (testnet) => async (txHash, vOut) => {
    const url = `${endpointV2(testnet)}/tx/${txHash}`;
    const response = await axios.get(`${url}`, {
        timeout: DEFAULT_TIMEOUT,
    });
    const utxo = response.data;
    return fixUTXO({
        txHash,
        amount: parseFloat(utxo.vout[vOut].value),
        // script_hex: utxo.scriptPubKey,
        vOut,
        confirmations: utxo.confirmations,
    }, 8);
};
const fetchUTXOs = (testnet) => async (address, confirmations) => {
    const url = `${endpointV2(testnet)}/addr/${address}/utxo`;
    const response = await axios.get(url, {
        timeout: DEFAULT_TIMEOUT,
    });
    return fixUTXOs(response.data
        .map((utxo) => ({
        txHash: utxo.txid,
        amount: utxo.amount,
        // script_hex: utxo.scriptPubKey,
        vOut: utxo.vout,
        confirmations: utxo.confirmations,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations), 8).sort(sortUTXOs);
};
const fetchTXs = (testnet) => async (address, confirmations) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/address/transactions/${address}`;
    const { data } = await axios.get(url, {
        timeout: DEFAULT_TIMEOUT,
    });
    const received = [];
    for (const tx of data.txs) {
        for (let i = 0; i < tx.vout.length; i++) {
            const vout = tx.vout[i];
            if (vout.scriptPubKey.addresses.indexOf(address) >= 0) {
                received.push({
                    txHash: tx.txid,
                    amount: fixValue(parseFloat(vout.value), 8),
                    vOut: i,
                    confirmations: tx.confirmations,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
export const broadcastTransaction = (testnet) => async (txHex) => {
    const url = `${endpoint(testnet).replace(/\/$/, "")}/rawtransactions/sendRawTransaction`;
    const response = await axios.post(url, { hexes: [txHex] }, { timeout: DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data[0];
};
export const BitcoinDotCom = {
    fetchUTXO,
    fetchUTXOs,
    fetchTXs,
    broadcastTransaction,
};
