// send-crypto/build/module/common/apis/blockchain.js
import axios from "axios";
import { URLSearchParams } from "url";
import { sortUTXOs } from "../../lib/utxo";
import { DEFAULT_TIMEOUT } from "./timeout";
export var BlockchainNetwork;
(function (BlockchainNetwork) {
    BlockchainNetwork["Bitcoin"] = "btc";
    BlockchainNetwork["BitcoinCash"] = "bch";
    BlockchainNetwork["BitcoinTestnet"] = "btc-testnet";
    BlockchainNetwork["BitcoinCashTestnet"] = "bch-testnet";
})(BlockchainNetwork || (BlockchainNetwork = {}));
const fetchLatestBlock = async (network) => {
    const statsUrl = `https://api.blockchain.info/haskoin-store/${network}/block/best?notx=true`;
    const statsResponse = (await axios.get(statsUrl)).data;
    return statsResponse.height;
};
const fetchUTXO = (network) => async (txHash, vOut) => {
    const url = `https://api.blockchain.info/haskoin-store/${network}/transaction/${txHash}`;
    const response = (await axios.get(`${url}`, {
        timeout: DEFAULT_TIMEOUT,
    })).data;
    const confirmations = !response.block || !response.block.height
        ? 0
        : Math.max((await fetchLatestBlock(network)) -
            response.block.height +
            1, 0);
    return {
        txHash,
        vOut,
        amount: response.outputs[vOut].value,
        confirmations,
    };
};
const fetchUTXOs = (network) => async (address, confirmations, limit = 25, offset = 0) => fetchTXs(network)(address, confirmations, limit, offset, true);
const fetchTXs = (network) => async (address, confirmations = 0, limit = 25, offset = 0, onlyUnspent = false) => {
    const url = `https://api.blockchain.info/haskoin-store/${network}/address/${address}/transactions/full?limit=${limit}&offset=${offset}`;
    const response = (await axios.get(url, {
        timeout: DEFAULT_TIMEOUT,
    })).data;
    let latestBlock;
    const received = [];
    for (const tx of response) {
        latestBlock = latestBlock || (await fetchLatestBlock(network));
        const txConfirmations = tx.block && tx.block.height
            ? Math.max(latestBlock - tx.block.height + 1, 0)
            : 0;
        for (let i = 0; i < tx.outputs.length; i++) {
            const vout = tx.outputs[i];
            if (vout.address === address &&
                // If the onlyUnspent flag is true, check that the tx is unspent.
                (!onlyUnspent || vout.spent === false)) {
                received.push({
                    txHash: tx.txid,
                    amount: vout.value,
                    vOut: i,
                    confirmations: txConfirmations,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
export const broadcastTransaction = (network) => async (txHex) => {
    if (network !== BlockchainNetwork.Bitcoin) {
        throw new Error(`Broadcasting ${network} transactions not supported by endpoint.`);
    }
    const url = `https://blockchain.info/pushtx`;
    const params = new URLSearchParams();
    params.append("tx", txHex);
    const response = await axios.post(url, params, {
        timeout: DEFAULT_TIMEOUT,
    });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    // TODO
    return response.data;
};
export const Blockchain = {
    networks: BlockchainNetwork,
    fetchUTXO,
    fetchUTXOs,
    broadcastTransaction,
    fetchTXs,
};
