// send-crypto/build/module/common/apis/electrumx.js
import axios from "axios";
import { sortUTXOs } from "../../lib/utxo";
const getURL = (network, testnet) => `https://multichain-web-proxy.herokuapp.com/electrumx-${network}-${testnet ? "testnet" : "mainnet"}`;
const fetchUTXOs = (network, testnet) => async (address, confirmations, scriptHash) => {
    if (!scriptHash) {
        throw new Error("Must provide script hash.");
    }
    const url = getURL(network, testnet);
    const latestBlock = (await axios.post(url, {
        jsonrpc: "1.0",
        id: "67",
        method: "blockchain.scripthash.listunspent",
        params: [scriptHash],
    })).data.result.height;
    const response = await axios.post(url, {
        jsonrpc: "1.0",
        id: "67",
        method: "blockchain.scripthash.listunspent",
        params: [scriptHash],
    });
    return response.data.result
        .map((utxo) => ({
        txHash: utxo.tx_hash,
        amount: utxo.value,
        vOut: utxo.tx_pos,
        confirmations: utxo.height ? 1 + latestBlock - utxo.height : 0,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(sortUTXOs);
};
export const ElectrumX = {
    fetchUTXOs,
};
