// @glif/filecoin-wallet-provider/dist/providers/local-node-provider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LocalNodeProvider = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _filecoinRpcClient = _interopRequireDefault(require("@glif/filecoin-rpc-client"));

var LocalNodeProvider = /*#__PURE__*/function () {
  function LocalNodeProvider(options) {
    (0, _classCallCheck2["default"])(this, LocalNodeProvider);
    (0, _defineProperty2["default"])(this, "apiAddress", void 0);
    (0, _defineProperty2["default"])(this, "token", void 0);
    (0, _defineProperty2["default"])(this, "jsonRpcEngine", void 0);
    this.apiAddress = options.apiAddress;
    this.token = options.token;
    this.jsonRpcEngine = new _filecoinRpcClient["default"]({
      apiAddress: options.apiAddress,
      token: options.token
    });
  }

  (0, _createClass2["default"])(LocalNodeProvider, [{
    key: "newAccount",
    value: function newAccount() {
      return this.jsonRpcEngine.request('WalletNew', 'secp256k1');
    }
  }, {
    key: "getAccounts",
    value: function getAccounts() {
      return this.jsonRpcEngine.request('WalletList');
    }
  }, {
    key: "sign",
    value: function sign(from, message) {
      return this.jsonRpcEngine.request('WalletSignMessage', from, message);
    }
  }]);
  return LocalNodeProvider;
}();

exports.LocalNodeProvider = LocalNodeProvider;
