// send-crypto/build/main/handlers/ETH/ETHHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ETHHandler = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const promiEvent_1 = require("../../lib/promiEvent");
const ethUtils_1 = require("./ethUtils");
class ETHHandler {
    constructor(privateKey, network, options, sharedState) {
        this.decimals = 18;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["ETH", "ETHER", "ETHEREUM"].indexOf(asset.toUpperCase()) !== -1;
        this.address = async (asset, options) => this.unlockedAddress;
        // Balance
        this.getBalance = async (asset, options) => (await this.getBalanceInSats(asset, options)).dividedBy(new bignumber_js_1.default(10).exponentiatedBy(this.decimals));
        this.getBalanceInSats = async (asset, options) => {
            let atBlock;
            if (options && options.confirmations && options.confirmations > 0) {
                const currentBlock = new bignumber_js_1.default(await this.sharedState.ethSigner.provider.getBlockNumber());
                atBlock = currentBlock
                    .minus(options.confirmations)
                    .plus(1)
                    .toNumber();
            }
            const address = (options && options.address) || (await this.address(asset));
            return new bignumber_js_1.default((await this.sharedState.ethSigner.provider.getBalance(address, atBlock)).toString());
        };
        // Transfer
        this.send = (to, value, asset, options) => this.sendSats(to, value.times(new bignumber_js_1.default(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, valueIn, asset, optionsIn) => {
            const promiEvent = (0, promiEvent_1.newPromiEvent)();
            (async () => {
                const options = optionsIn || {};
                let value = valueIn;
                const txOptions = (0, ethUtils_1.getTransactionConfig)(options);
                if (options.subtractFee) {
                    const gasPrice = txOptions.gasPrice ||
                        (await this.sharedState.ethSigner.provider.getGasPrice());
                    const gasPriceBN = new bignumber_js_1.default(gasPrice.toString());
                    const gasLimit = txOptions.gasLimit || 21000;
                    const gasLimitBN = new bignumber_js_1.default(gasLimit.toString());
                    const fee = gasPriceBN.times(gasLimitBN);
                    if (fee.gt(value)) {
                        throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                    }
                    value = value.minus(fee);
                }
                const from = await this.address(asset);
                const tx = await this.sharedState.ethSigner.sendTransaction(Object.assign(Object.assign({ from, gasLimit: 21000 }, txOptions), { to, value: value.toFixed() }));
                promiEvent.emit("transactionHash", tx.hash);
                await tx.wait();
                promiEvent.resolve(tx.hash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.network = (0, ethUtils_1.getNetwork)(network);
        this.privateKey = privateKey;
        const [ethSigner, address] = (0, ethUtils_1.getEthersSigner)(this.privateKey, (0, ethUtils_1.getEndpoint)(this.network, options && options.ethereumNode, options && options.infuraKey));
        this.unlockedAddress = address;
        sharedState.ethSigner = ethSigner;
        this.sharedState = sharedState;
    }
}
exports.ETHHandler = ETHHandler;
