// send-crypto/build/main/handlers/ETH/ethUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransactionConfig = exports.getNetwork = exports.Network = exports.getEndpoint = exports.getEthersSigner = void 0;
const ethers_1 = require("ethers");
const getEthersSigner = (privateKey, endpoint) => {
    // const provider = new HDWalletProvider(privateKey, endpoint);
    const provider = new ethers_1.ethers.providers.JsonRpcProvider(endpoint);
    const signer = new ethers_1.ethers.Wallet(privateKey, provider);
    return [signer, signer.address];
};
exports.getEthersSigner = getEthersSigner;
// Free tier - only used as a fallback.
const defaultInfuraKey = "3b7a6c29f9c048d688a848899888aa96";
const getEndpoint = (network, ethereumNode, infuraKey) => {
    return ethereumNode
        ? ethereumNode
        : `https://${network}.infura.io/v3/${infuraKey || defaultInfuraKey}`;
};
exports.getEndpoint = getEndpoint;
var Network;
(function (Network) {
    Network["Mainnet"] = "mainnet";
    Network["Ropsten"] = "ropsten";
    Network["Kovan"] = "kovan";
    Network["Rinkeby"] = "rinkeby";
    Network["G\u00F6rli"] = "goerli";
})(Network = exports.Network || (exports.Network = {}));
const getNetwork = (network) => {
    switch (network.toLowerCase()) {
        case "mainnet":
        case "main":
            return Network.Mainnet;
        case "kovan":
            return Network.Kovan;
        case "rinkeby":
            return Network.Rinkeby;
        case "görli":
        case "goerli":
        case "gorli":
            return Network.Görli;
        case "ropsten":
        case "testnet":
        default:
            return Network.Ropsten;
    }
};
exports.getNetwork = getNetwork;
// Create a `txConfig` object with only the relevant fields in the `options`
// object.
const getTransactionConfig = (options) => {
    const txConfig = {};
    // tslint:disable: no-object-mutation
    if (options.gasLimit) {
        txConfig.gasLimit = options.gasLimit;
    }
    if (options.gasPrice) {
        txConfig.gasPrice = options.gasPrice;
    }
    if (options.maxFeePerGas) {
        txConfig.maxFeePerGas = options.maxFeePerGas;
    }
    if (options.maxPriorityFeePerGas) {
        txConfig.maxPriorityFeePerGas = options.maxPriorityFeePerGas;
    }
    if (options.nonce) {
        txConfig.nonce = options.nonce;
    }
    if (options.type) {
        txConfig.type = options.type;
    }
    if (options.accessList) {
        txConfig.accessList = options.accessList;
    }
    if (options.customData) {
        txConfig.customData = options.customData;
    }
    return txConfig;
};
exports.getTransactionConfig = getTransactionConfig;
