// send-crypto/build/main/common/apis/blockchain.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blockchain = exports.broadcastTransaction = exports.BlockchainNetwork = void 0;
const axios_1 = __importDefault(require("axios"));
const url_1 = require("url");
const utxo_1 = require("../../lib/utxo");
const timeout_1 = require("./timeout");
var BlockchainNetwork;
(function (BlockchainNetwork) {
    BlockchainNetwork["Bitcoin"] = "btc";
    BlockchainNetwork["BitcoinCash"] = "bch";
    BlockchainNetwork["BitcoinTestnet"] = "btc-testnet";
    BlockchainNetwork["BitcoinCashTestnet"] = "bch-testnet";
})(BlockchainNetwork = exports.BlockchainNetwork || (exports.BlockchainNetwork = {}));
const fetchLatestBlock = async (network) => {
    const statsUrl = `https://api.blockchain.info/haskoin-store/${network}/block/best?notx=true`;
    const statsResponse = (await axios_1.default.get(statsUrl)).data;
    return statsResponse.height;
};
const fetchUTXO = (network) => async (txHash, vOut) => {
    const url = `https://api.blockchain.info/haskoin-store/${network}/transaction/${txHash}`;
    const response = (await axios_1.default.get(`${url}`, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    })).data;
    const confirmations = !response.block || !response.block.height
        ? 0
        : Math.max((await fetchLatestBlock(network)) -
            response.block.height +
            1, 0);
    return {
        txHash,
        vOut,
        amount: response.outputs[vOut].value,
        confirmations,
    };
};
const fetchUTXOs = (network) => async (address, confirmations, limit = 25, offset = 0) => fetchTXs(network)(address, confirmations, limit, offset, true);
const fetchTXs = (network) => async (address, confirmations = 0, limit = 25, offset = 0, onlyUnspent = false) => {
    const url = `https://api.blockchain.info/haskoin-store/${network}/address/${address}/transactions/full?limit=${limit}&offset=${offset}`;
    const response = (await axios_1.default.get(url, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    })).data;
    let latestBlock;
    const received = [];
    for (const tx of response) {
        latestBlock = latestBlock || (await fetchLatestBlock(network));
        const txConfirmations = tx.block && tx.block.height
            ? Math.max(latestBlock - tx.block.height + 1, 0)
            : 0;
        for (let i = 0; i < tx.outputs.length; i++) {
            const vout = tx.outputs[i];
            if (vout.address === address &&
                // If the onlyUnspent flag is true, check that the tx is unspent.
                (!onlyUnspent || vout.spent === false)) {
                received.push({
                    txHash: tx.txid,
                    amount: vout.value,
                    vOut: i,
                    confirmations: txConfirmations,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const broadcastTransaction = (network) => async (txHex) => {
    if (network !== BlockchainNetwork.Bitcoin) {
        throw new Error(`Broadcasting ${network} transactions not supported by endpoint.`);
    }
    const url = `https://blockchain.info/pushtx`;
    const params = new url_1.URLSearchParams();
    params.append("tx", txHex);
    const response = await axios_1.default.post(url, params, {
        timeout: timeout_1.DEFAULT_TIMEOUT,
    });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    // TODO
    return response.data;
};
exports.broadcastTransaction = broadcastTransaction;
exports.Blockchain = {
    networks: BlockchainNetwork,
    fetchUTXO,
    fetchUTXOs,
    broadcastTransaction: exports.broadcastTransaction,
    fetchTXs,
};
