// send-crypto/build/main/common/apis/electrumx.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElectrumX = void 0;
const axios_1 = __importDefault(require("axios"));
const utxo_1 = require("../../lib/utxo");
const getURL = (network, testnet) => `https://multichain-web-proxy.herokuapp.com/electrumx-${network}-${testnet ? "testnet" : "mainnet"}`;
const fetchUTXOs = (network, testnet) => async (address, confirmations, scriptHash) => {
    if (!scriptHash) {
        throw new Error("Must provide script hash.");
    }
    const url = getURL(network, testnet);
    const latestBlock = (await axios_1.default.post(url, {
        jsonrpc: "1.0",
        id: "67",
        method: "blockchain.scripthash.listunspent",
        params: [scriptHash],
    })).data.result.height;
    const response = await axios_1.default.post(url, {
        jsonrpc: "1.0",
        id: "67",
        method: "blockchain.scripthash.listunspent",
        params: [scriptHash],
    });
    return response.data.result
        .map((utxo) => ({
        txHash: utxo.tx_hash,
        amount: utxo.value,
        vOut: utxo.tx_pos,
        confirmations: utxo.height ? 1 + latestBlock - utxo.height : 0,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
exports.ElectrumX = {
    fetchUTXOs,
};
