// send-crypto/build/main/common/apis/insight.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Insight = exports.broadcastTransaction = void 0;
const axios_1 = __importDefault(require("axios"));
const https_1 = __importDefault(require("https"));
const utxo_1 = require("../../lib/utxo");
const timeout_1 = require("./timeout");
const fetchUTXOs = (insightURL) => async (address, confirmations) => {
    const url = `${insightURL.replace(/\/$/, "")}/addr/${address}/utxo`;
    const response = await axios_1.default.get(url, {
        // TODO: Remove when certificate is fixed.
        httpsAgent: new https_1.default.Agent({
            rejectUnauthorized: false,
        }),
        timeout: timeout_1.DEFAULT_TIMEOUT,
    });
    const data = typeof response.data === "string"
        ? JSON.parse(response.data)
        : response.data;
    return data
        .map((utxo) => ({
        txHash: utxo.txid,
        amount: utxo.satoshis || (0, utxo_1.fixValue)(utxo.amount, 8),
        // script_hex: utxo.scriptPubKey,
        vOut: utxo.vout,
        confirmations: utxo.confirmations,
    }))
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const fetchTXs = (insightURL) => async (address, confirmations) => {
    const url = `${insightURL.replace(/\/$/, "")}/txs/?address=${address}`;
    const response = await axios_1.default.get(url, {
        // TODO: Remove when certificate is fixed.
        httpsAgent: new https_1.default.Agent({
            rejectUnauthorized: false,
        }),
        timeout: timeout_1.DEFAULT_TIMEOUT,
    });
    const data = typeof response.data === "string"
        ? JSON.parse(response.data)
        : response.data;
    const received = [];
    for (const tx of data.txs) {
        for (let i = 0; i < tx.vout.length; i++) {
            const vout = tx.vout[i];
            if (vout.scriptPubKey.addresses.indexOf(address) >= 0) {
                received.push({
                    txHash: tx.txid,
                    amount: (0, utxo_1.fixValue)(parseFloat(vout.value), 8),
                    vOut: i,
                    confirmations: tx.confirmations,
                });
            }
        }
    }
    return received
        .filter((utxo) => confirmations === 0 || utxo.confirmations >= confirmations)
        .sort(utxo_1.sortUTXOs);
};
const fetchUTXO = (insightURL) => async (txHash, vOut) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/${txHash}`;
    const tx = (await axios_1.default.get(url, { timeout: timeout_1.DEFAULT_TIMEOUT }))
        .data;
    return (0, utxo_1.fixUTXO)({
        txHash,
        amount: parseFloat(tx.vout[vOut].value),
        vOut,
        confirmations: tx.confirmations,
    }, 8);
};
const broadcastTransaction = (insightURL) => async (txHex) => {
    const url = `${insightURL.replace(/\/$/, "")}/tx/send`;
    const response = await axios_1.default.post(url, { rawtx: txHex }, { timeout: timeout_1.DEFAULT_TIMEOUT });
    if (response.data.error) {
        throw new Error(response.data.error);
    }
    return response.data.txid;
};
exports.broadcastTransaction = broadcastTransaction;
exports.Insight = {
    fetchUTXO,
    fetchUTXOs,
    fetchTXs,
    broadcastTransaction: exports.broadcastTransaction,
};
