// send-crypto/build/main/common/apis/jsonrpc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONRPC = exports.MULTICHAIN_URLS = void 0;
const axios_1 = __importDefault(require("axios"));
const timeout_1 = require("./timeout");
exports.MULTICHAIN_URLS = {
    BTC: "https://multichain-web-proxy.herokuapp.com/multichain-bitcoin",
    BTCTEST: "https://multichain-web-proxy.herokuapp.com/multichain-bitcoin-testnet",
    ZEC: "https://multichain-web-proxy.herokuapp.com/multichain-zcash",
    ZECTEST: "https://multichain-web-proxy.herokuapp.com/multichain-zcash-testnet",
    BCH: "https://multichain-web-proxy.herokuapp.com/multichain-bitcoincash",
    BCHTEST: "https://multichain-web-proxy.herokuapp.com/multichain-bitcoincash-testnet",
};
const broadcastTransaction = (url) => async (txHex) => {
    const response = await axios_1.default.post(url, {
        jsonrpc: "1.0",
        id: "67",
        method: "sendrawtransaction",
        params: [txHex],
    }, { timeout: timeout_1.DEFAULT_TIMEOUT });
    return response.data.result;
};
exports.JSONRPC = {
    broadcastTransaction,
};
