// send-crypto/build/main/handlers/TERRA/TERRAHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TERRAHandler = exports.TerraNetwork = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const terra_js_1 = require("@terra-money/terra.js");
const promiEvent_1 = require("../../lib/promiEvent");
const utils_1 = require("../../lib/utils");
var TerraNetwork;
(function (TerraNetwork) {
    TerraNetwork["Tequila"] = "tequila-0004";
    TerraNetwork["Columbus"] = "columbus-3";
})(TerraNetwork = exports.TerraNetwork || (exports.TerraNetwork = {}));
const toDenom = (asset) => "uluna";
class TERRAHandler {
    constructor(privateKey, network, options = {}, sharedState) {
        this.decimals = 6;
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => typeof asset === "string" &&
            ["LUNA"].indexOf(asset.toUpperCase()) !== -1;
        this.address = (asset, _options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            return this.key.accAddress;
        };
        // Balance
        this.getBalance = async (asset, options = {}) => (await this.getBalanceInSats(asset, options)).dividedBy(new bignumber_js_1.default(10).exponentiatedBy(this.decimals));
        this.getBalanceInSats = async (asset, options = {}) => {
            if (!this.handlesAsset(asset)) {
                throw new Error(`Asset ${asset} not supported.`);
            }
            const balances = (await this.client.bank.balance((options && options.address) || (await this.address(asset))))[0];
            const balance = balances.get(toDenom(asset));
            return new bignumber_js_1.default(balance ? balance.amount.toFixed() : 0);
        };
        // Transfer
        this.send = (to, value, asset, options = {}) => this.sendSats(to, value.times(new bignumber_js_1.default(10).exponentiatedBy(this.decimals)), asset, options);
        this.sendSats = (to, value, asset, options = {}) => {
            const promiEvent = (0, promiEvent_1.newPromiEvent)();
            (async () => {
                const address = await this.address(asset);
                const send = new terra_js_1.MsgSend(address, to, { uluna: value.toFixed() });
                const signedTx = await this.wallet.createAndSignTx({
                    msgs: [send],
                    memo: options.memo,
                });
                const result = await this.client.tx.broadcast(signedTx);
                const txHash = result.txhash;
                promiEvent.emit("transactionHash", txHash);
                promiEvent.resolve(txHash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.network =
            network === "mainnet"
                ? TerraNetwork.Columbus
                : TerraNetwork.Tequila;
        const client = sharedState.client ||
            new terra_js_1.LCDClient({
                URL: options && options.terra ? options.terra.URL : "",
                chainID: this.network,
            });
        const key = new terra_js_1.RawKey(Buffer.from((0, utils_1.strip0x)(privateKey), "hex"));
        const wallet = client.wallet(key);
        this.client = client;
        this.wallet = wallet;
        this.key = key;
    }
}
exports.TERRAHandler = TERRAHandler;
