// send-crypto/build/module/handlers/ETH/ETHHandler.js
import BigNumber from "bignumber.js";
import { newPromiEvent } from "../../lib/promiEvent";
import { getEndpoint, getEthersSigner, getNetwork, getTransactionConfig, } from "./ethUtils";
export class ETHHandler {
    privateKey;
    network;
    decimals = 18;
    unlockedAddress;
    sharedState;
    constructor(privateKey, network, options, sharedState) {
        this.network = getNetwork(network);
        this.privateKey = privateKey;
        const [ethSigner, address] = getEthersSigner(this.privateKey, getEndpoint(this.network, options && options.ethereumNode, options && options.infuraKey));
        this.unlockedAddress = address;
        sharedState.ethSigner = ethSigner;
        this.sharedState = sharedState;
    }
    // Returns whether or not this can handle the asset
    handlesAsset = (asset) => typeof asset === "string" &&
        ["ETH", "ETHER", "ETHEREUM"].indexOf(asset.toUpperCase()) !== -1;
    address = async (asset, options) => this.unlockedAddress;
    // Balance
    getBalance = async (asset, options) => (await this.getBalanceInSats(asset, options)).dividedBy(new BigNumber(10).exponentiatedBy(this.decimals));
    getBalanceInSats = async (asset, options) => {
        let atBlock;
        if (options && options.confirmations && options.confirmations > 0) {
            const currentBlock = new BigNumber(await this.sharedState.ethSigner.provider.getBlockNumber());
            atBlock = currentBlock
                .minus(options.confirmations)
                .plus(1)
                .toNumber();
        }
        const address = (options && options.address) || (await this.address(asset));
        return new BigNumber((await this.sharedState.ethSigner.provider.getBalance(address, atBlock)).toString());
    };
    // Transfer
    send = (to, value, asset, options) => this.sendSats(to, value.times(new BigNumber(10).exponentiatedBy(this.decimals)), asset, options);
    sendSats = (to, valueIn, asset, optionsIn) => {
        const promiEvent = newPromiEvent();
        (async () => {
            const options = optionsIn || {};
            let value = valueIn;
            const txOptions = getTransactionConfig(options);
            if (options.subtractFee) {
                const gasPrice = txOptions.gasPrice ||
                    (await this.sharedState.ethSigner.provider.getGasPrice());
                const gasPriceBN = new BigNumber(gasPrice.toString());
                const gasLimit = txOptions.gasLimit || 21000;
                const gasLimitBN = new BigNumber(gasLimit.toString());
                const fee = gasPriceBN.times(gasLimitBN);
                if (fee.gt(value)) {
                    throw new Error(`Unable to include fee in value, fee exceeds value (${fee.toFixed()} > ${value.toFixed()})`);
                }
                value = value.minus(fee);
            }
            const from = await this.address(asset);
            const tx = await this.sharedState.ethSigner.sendTransaction({
                from,
                gasLimit: 21000,
                ...txOptions,
                to,
                value: value.toFixed(),
            });
            promiEvent.emit("transactionHash", tx.hash);
            await tx.wait();
            promiEvent.resolve(tx.hash);
        })().catch((error) => {
            promiEvent.reject(error);
        });
        return promiEvent;
    };
}
