// send-crypto/build/main/handlers/ERC20/ERC20Handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC20Handler = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const ethers_1 = require("ethers");
const promiEvent_1 = require("../../lib/promiEvent");
const ethUtils_1 = require("../ETH/ethUtils");
const ERC20ABI_1 = require("./ERC20ABI");
const ERC20s_1 = require("./ERC20s");
const resolveAsset = (network, assetIn) => {
    if (typeof assetIn !== "object") {
        throw new Error("");
    }
    const asset = assetIn;
    if (asset.address) {
        return Object.assign(Object.assign({}, asset), { address: asset.address });
    }
    else {
        const address = (ERC20s_1.ERC20s[network] || {})[asset.name || ""];
        if (!address) {
            throw new Error(`Unknown ERC20 token ${asset.name || JSON.stringify(asset)}`);
        }
        return Object.assign(Object.assign({}, asset), { address });
    }
};
class ERC20Handler {
    constructor(_privateKey, network, _options, sharedState) {
        this._decimals = {};
        // Returns whether or not this can handle the asset
        this.handlesAsset = (asset) => {
            return (typeof asset === "object" &&
                asset.hasOwnProperty("type") &&
                asset.type === "ERC20" &&
                (asset.hasOwnProperty("address") || asset.hasOwnProperty("name")));
        };
        // Address
        this.address = async (asset, options, deferHandler) => deferHandler.address("ETH", options);
        // Balance
        this.getBalance = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const decimals = await this.decimals(asset);
            return (await this.getBalanceInSats(asset, options, deferHandler)).dividedBy(new bignumber_js_1.default(10).exponentiatedBy(decimals));
        };
        this.getBalanceInSats = async (assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const address = (options && options.address) ||
                (deferHandler && (await deferHandler.address("ETH", options))) ||
                "";
            return new bignumber_js_1.default((await this.getContract(asset).balanceOf(address)).toString());
        };
        // Transfer
        // This is re-implemented instead of calling sendSats so that a PromiEvent
        // can be returned.
        this.send = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = (0, promiEvent_1.newPromiEvent)();
            (async () => {
                const value = valueIn.times(new bignumber_js_1.default(10).exponentiatedBy(await this.decimals(asset)));
                const contract = this.getContract(asset);
                const config = Object.assign({ from: await deferHandler.address("ETH") }, (0, ethUtils_1.getTransactionConfig)(options));
                let tx;
                // tslint:disable: prefer-conditional-expression
                if (options.approve) {
                    // config.gasLimit = contract.estimateGas.approve(to, valueIn.toFixed());
                    tx = await contract.approve(to, value.toFixed(), config);
                }
                else {
                    // config.gasLimit = contract.estimateGas.transfer(to, valueIn.toFixed());
                    tx = await contract.transfer(to, value.toFixed(), config);
                }
                promiEvent.emit("transactionHash", tx.hash);
                await tx.wait();
                promiEvent.resolve(tx.hash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.sendSats = (to, valueIn, assetIn, options, deferHandler) => {
            const asset = resolveAsset(this.network, assetIn);
            const promiEvent = (0, promiEvent_1.newPromiEvent)();
            (async () => {
                const contract = this.getContract(asset);
                const config = Object.assign({ from: await deferHandler.address("ETH") }, (0, ethUtils_1.getTransactionConfig)(options));
                // tslint:disable: prefer-conditional-expression
                let tx;
                if (options.approve) {
                    // config.gasLimit = contract.estimateGas.approve(to, valueIn.toFixed());
                    tx = await contract.approve(to, valueIn.toFixed(), config);
                }
                else {
                    // config.gasLimit = contract.estimateGas.transfer(to, valueIn.toFixed());
                    tx = await contract.transfer(to, valueIn.toFixed(), config);
                }
                promiEvent.emit("transactionHash", tx.hash);
                await tx.wait();
                promiEvent.resolve(tx.hash);
            })().catch((error) => {
                promiEvent.reject(error);
            });
            return promiEvent;
        };
        this.getContract = (asset) => {
            return new ethers_1.ethers.Contract(resolveAsset(this.network, asset).address, ERC20ABI_1.ERC20ABI, this.sharedState.ethSigner);
        };
        this.decimals = async (asset) => {
            const address = resolveAsset(this.network, asset).address;
            if (this._decimals[address]) {
                return this._decimals[address];
            }
            return new bignumber_js_1.default((await this.getContract(asset).decimals()).toString()).toNumber();
        };
        this.network = (0, ethUtils_1.getNetwork)(network);
        this.sharedState = sharedState;
    }
}
exports.ERC20Handler = ERC20Handler;
