// @glif/filecoin-number/dist/FilecoinNumber.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilecoinNumber = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _bignumber = require("bignumber.js");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

// not sure how we want to configure rounding for this
_bignumber.BigNumber.set({
  ROUNDING_MODE: _bignumber.BigNumber.ROUND_HALF_DOWN
});

_bignumber.BigNumber.config({
  EXPONENTIAL_AT: 1e9
});

function asBigNumber(amount, denom) {
  if (!denom) {
    throw new Error('No Filecoin denomination passed in constructor.');
  }

  var formattedDenom = denom.toLowerCase();

  if (formattedDenom !== 'fil' && formattedDenom !== 'picofil' && formattedDenom !== 'attofil') {
    throw new Error('Unsupported denomination passed in constructor. Must pass picofil or attofil.');
  }

  if (formattedDenom === 'picofil') {
    return new _bignumber.BigNumber(amount).shiftedBy(-12);
  } else if (formattedDenom === 'attofil') {
    return new _bignumber.BigNumber(amount).shiftedBy(-18);
  } else {
    return amount;
  }
} // stores filecoin numbers in denominations of Fil, not AttoFil


var FilecoinNumber = /*#__PURE__*/function (_BigNumber) {
  (0, _inherits2["default"])(FilecoinNumber, _BigNumber);

  var _super = _createSuper(FilecoinNumber);

  function FilecoinNumber(amount, denom) {
    (0, _classCallCheck2["default"])(this, FilecoinNumber);
    return _super.call(this, asBigNumber(amount, denom));
  }

  (0, _createClass2["default"])(FilecoinNumber, [{
    key: "toFil",
    value: function toFil() {
      return this.toString();
    }
  }, {
    key: "toPicoFil",
    value: function toPicoFil() {
      return this.shiftedBy(12).toString();
    }
  }, {
    key: "toAttoFil",
    value: function toAttoFil() {
      return this.shiftedBy(18).toFixed(0, 1);
    }
  }]);
  return FilecoinNumber;
}(_bignumber.BigNumber);

exports.FilecoinNumber = FilecoinNumber;
